# 第三周课程内容：python基础（函数）至Pandas基础篇

## 简介和基础知识
经过第三周的学习，相信大家对python中的函数应用和Numpy、Pandas两大数据分析模块的基础应用已经有了一定的了解。

本周的挑战任务会分为**两部分**：
* 普通的代码题
* 综合应用大题

让我们来实际应用一下所学到的知识和方法，做一个简单的数据处理和分析。


### 代码题
### 1.
```
a,b = 5,8
c = a**b
```
### 将以上代码改为函数（输入变量为a和b），函数名为calc，并调用函数，打印运行结果：

**（示例）**
```
def calc(a, b):
    
    '''你的代码'''
    
    return     # 结果
    
print()  # 打印结果
```

In [1]:
'''请输入你的代码'''
def calc(a, b):
    return a**b
print(calc(5,8))

390625


### 2.	现有数组x = np.arange(9.).reshape(3, 3)，利用np.where函数，使得数组中大于4的值为1，小于等于4的值为0，代码为：

In [92]:
'''请输入你的代码'''
# 这里需要注意顺序，先小于等于 再大于
# where 返回的是数组位置，并不是值

import numpy as np
x = np.arange(9.).reshape(3, 3)
x[np.where(x <= 4)] = 0
x[np.where(x > 4)] = 1

[[0. 1. 2.]
 [3. 4. 5.]
 [6. 7. 8.]]
[[0. 0. 0.]
 [0. 0. 5.]
 [6. 7. 8.]]
[[0. 0. 0.]
 [0. 0. 1.]
 [1. 1. 1.]]


## 应用题
现在我们手上有一份客户数据，内容是一些客户的基本资料、投资的账户、投资的收益以及账户的余额。这些数据都保存在Investors.xlsx中。
请下载Investors.xlsx，与本Notebook文件保存在同一文件夹下。

### 开始
要开始处理客户数据，我们首先需要  `import` 所需的功能，并将数据导入 pandas。

利用DataFrame中的函数运行以下代码单元格，加载我们的数据并使用 `.head(15) ` 函数显示前几个客户以进行检查。
> **提示：**你可以通过点击单元格和使用键盘快捷键 **Shift + Enter** 或 **Shift + Return** 运行代码单元格。此外，在选中单元格后还可以使用工具栏中的**播放**按钮执行代码单元格。双击 Markdown 单元格（如下所示的文本单元格）即可进行编辑，并使用相同的快捷键保存。

In [18]:
#Import Numpy and Pandas
import numpy as np
import pandas as pd
from IPython.display import display   # Allows the use of display() for DataFrames

# Pretty display for notebooks
%matplotlib inline

# TO DO: Load the data to 'full_data'
full_data = pd.read_excel('Investors.xlsx')

# Print the first few rows of the data
display(full_data.head(15))

姓名 性别  年龄 投资市场    投资账户       开户时间      投资收益       资产余额
0    张伟  男  32   A股  A00001 2016-03-02  30167.47  152582.83
1    王伟  男  26   A股  A00002 2016-06-04  87671.93  106538.72
2    王芳  女  50   A股  A00003 2017-04-21 -18090.58  125266.82
3    王芳  女  50   港股  H00003 2018-06-12  -7323.56  190558.09
4    李伟  男  31   A股  A00004 2016-09-01  27009.14  195731.90
5    李娜  女  58   美股  U00004 2018-04-24 -19278.16  132835.86
6   王秀英  女  56   A股  A00005 2017-07-05  40358.34  140649.78
7   李秀英  女  45   A股  A00006 2018-06-23  36510.09  115750.13
8   张秀英  女  52   A股  A00008 2017-11-03   6220.14  116819.85
9    刘伟  男  46   A股  A00009 2017-10-21   7210.51  168387.54
10   张敏  女  45   A股  A00010 2016-09-13  -3225.00  176235.81
11   李伟  男  29   A股  A00011 2018-03-24  42278.11  159249.38
12  李秀英  女  28   A股  A00012 2018-03-05  43597.13  121773.59
13   张伟  男  32   港股  H00001 2017-02-14   2026.84  178320.86
14   王伟  男  26   港股  H00002 2016-06-26  54454.37  107915.45

让我们来仔细看一下，这个客户数据表格的内容：
- **姓名：** 客户的中文姓名。有重名，例如编号4和编号编号11，姓名相同，但是年龄不同，账号不同，是两个不同的人
- **性别：** 客户的性别（男/女）
- **年龄：** 客户的年龄
- **投资市场：** 客户投资的市场（A股/港股/美股）
- **投资账户号码：** 客户的账户号码，其中A-A股，H-港股，U-美股
- **开户时间：** 客户开户的具体日期
- **投资收益：** 表示开户至今（2018年9月1日），客户的累计收益/损失
- **资产余额：** 表示客户在本时点（2018年9月1日），账户中所有资产的余额

## 一些描述性的统计

### 问题1（4分）：
我们先来尝试回答这几个简单的问题：
1. 投资者中，年龄最大的是几岁？
2. 开户时间最早的是哪一天开户的？
3. 所有投资者的资产余额，平均数是多少？中位数又是多少？（保留2位小数）

>**提示：** numpy中有min, max等常用的函数可以使用。 print 资产余额时注意调整打印格式(可以参考Week2挑战题第一题所提到的方法)

In [19]:
# 将下列''替换为你的代码
# TO DO: Print Largest Age among all investors
largest_age = np.max(full_data['年龄'])

# TO DO: Print Earliest Account Open Date
earliest_open_date = np.min(full_data['开户时间'])

# TO DO: Print Average & Median of Account Balance
average_account_balance = np.mean(full_data['资产余额'])
median_account_balance = np.median(full_data['资产余额'])

# Print result
print('Largest Age among all investors is', largest_age)
print('Earliest Account Open Date is', earliest_open_date)
print('Average Account Balance is', "%.2f" % average_account_balance)
print('Median Account Balance is', "%.2f" % median_account_balance)

Largest Age among all investors is 58
Earliest Account Open Date is 2016-01-17 00:00:00
Average Account Balance is 146886.05
Median Account Balance is 140649.78


**最后打印结果应该为**
- Largest Age among all investors is 58
- Earliest Account Open Date is 2016-01-17 00:00:00
- Average Account Balance is 146886.05
- Median Account Balance is 140649.78

除了以上的方法，我们还可以直接用describe来查看数据的统计性质。
请使用describe来查看，并自行检验上一步骤的运算结果是否正确。

In [25]:
# TO DO: Print Description of full_data’
print(full_data['年龄'].describe())
print(full_data['开户时间'].describe())
print(full_data['资产余额'].describe())

count    25.000000
mean     41.360000
std      11.434159
min      26.000000
25%      31.000000
50%      45.000000
75%      50.000000
max      58.000000
Name: 年龄, dtype: float64
count                      25
unique                     24
top       2018-06-23 00:00:00
freq                        2
first     2016-01-17 00:00:00
last      2018-06-23 00:00:00
Name: 开户时间, dtype: object
count        25.000000
mean     146886.045200
std       29134.270837
min      106538.720000
25%      121773.590000
50%      140649.780000
75%      175032.340000
max      195731.900000
Name: 资产余额, dtype: float64


## 进一步的处理
我们已经知道，重名的投资者很多。我们现在来尝试创建一个新的列作为唯一的key

### 问题2（1分）：
创建新的一列'new'，这一列的值为 姓名 + 投资账户，要求更新原full_data，然后用display和`head()`命令，显示数据的前几行

In [29]:
# TO DO: Create 'new' column, Value = Name + A/C No.
'''请输入你的代码'''
full_data['new'] = full_data['姓名'] + full_data['投资账户']
display(full_data.head(15))

0      张伟A00001
1      王伟A00002
2      王芳A00003
3      王芳H00003
4      李伟A00004
5      李娜U00004
6     王秀英A00005
7     李秀英A00006
8     张秀英A00008
9      刘伟A00009
10     张敏A00010
11     李伟A00011
12    李秀英A00012
13     张伟H00001
14     王伟H00002
15    李秀英H00004
16     李娜H00005
17    张秀英H00006
18     刘伟H00007
19     李娜H00008
20    张秀英H00009
21     张伟U00001
22     王伟U00002
23    李秀英U00003
24     李娜A00007
Name: new, dtype: object


姓名 性别  年龄 投资市场    投资账户       开户时间      投资收益       资产余额        new
0    张伟  男  32   A股  A00001 2016-03-02  30167.47  152582.83   张伟A00001
1    王伟  男  26   A股  A00002 2016-06-04  87671.93  106538.72   王伟A00002
2    王芳  女  50   A股  A00003 2017-04-21 -18090.58  125266.82   王芳A00003
3    王芳  女  50   港股  H00003 2018-06-12  -7323.56  190558.09   王芳H00003
4    李伟  男  31   A股  A00004 2016-09-01  27009.14  195731.90   李伟A00004
5    李娜  女  58   美股  U00004 2018-04-24 -19278.16  132835.86   李娜U00004
6   王秀英  女  56   A股  A00005 2017-07-05  40358.34  140649.78  王秀英A00005
7   李秀英  女  45   A股  A00006 2018-06-23  36510.09  115750.13  李秀英A00006
8   张秀英  女  52   A股  A00008 2017-11-03   6220.14  116819.85  张秀英A00008
9    刘伟  男  46   A股  A00009 2017-10-21   7210.51  168387.54   刘伟A00009
10   张敏  女  45   A股  A00010 2016-09-13  -3225.00  176235.81   张敏A00010
11   李伟  男  29   A股  A00011 2018-03-24  42278.11  159249.38   李伟A00011
12  李秀英  女  28   A股  A00012 2018-03-05  43597.13  121773.59  李秀英A00012
13   张伟  男  32   港股  H00001 2017-02-14   2026.84  178320.86   张伟H00001
14   王伟  男  26   港股  H00002 2016-06-26  54454.37  107915.45   王伟H00002

### 问题3（2分）：
- 创建新的一列'Aging'，这一列的值为：报告日期(report_date: 2018-09-01)减去开户时间，表示该客户开户至今已有多长时间。
- 要求更新原full_data，然后用display和`head()`命令，显示数据的前几行

In [32]:
import datetime
report_date = datetime.datetime.strptime('2018-09-01', '%Y-%m-%d') #对report_date进行数据格式转换

# TO DO: Create 'Aging' column, Value = report_date - A/C Open Date
full_data['Aging'] = report_date - full_data['开户时间']
display(full_data.head(15))

姓名 性别  年龄 投资市场    投资账户       开户时间      投资收益       资产余额        new  \
0    张伟  男  32   A股  A00001 2016-03-02  30167.47  152582.83   张伟A00001   
1    王伟  男  26   A股  A00002 2016-06-04  87671.93  106538.72   王伟A00002   
2    王芳  女  50   A股  A00003 2017-04-21 -18090.58  125266.82   王芳A00003   
3    王芳  女  50   港股  H00003 2018-06-12  -7323.56  190558.09   王芳H00003   
4    李伟  男  31   A股  A00004 2016-09-01  27009.14  195731.90   李伟A00004   
5    李娜  女  58   美股  U00004 2018-04-24 -19278.16  132835.86   李娜U00004   
6   王秀英  女  56   A股  A00005 2017-07-05  40358.34  140649.78  王秀英A00005   
7   李秀英  女  45   A股  A00006 2018-06-23  36510.09  115750.13  李秀英A00006   
8   张秀英  女  52   A股  A00008 2017-11-03   6220.14  116819.85  张秀英A00008   
9    刘伟  男  46   A股  A00009 2017-10-21   7210.51  168387.54   刘伟A00009   
10   张敏  女  45   A股  A00010 2016-09-13  -3225.00  176235.81   张敏A00010   
11   李伟  男  29   A股  A00011 2018-03-24  42278.11  159249.38   李伟A00011   
12  李秀英  女  28   A股  A00012 2018-03-05  43597.13  121773.59  李秀英A00012   
13   张伟  男  32   港股  H00001 2017-02-14   2026.84  178320.86   张伟H00001   
14   王伟  男  26   港股  H00002 2016-06-26  54454.37  107915.45   王伟H00002   

      Aging  
0  913 days  
1  819 days  
2  498 days  
3   81 days  
4  730 days  
5  130 days  
6  423 days  
7   70 days  
8  302 days  
9  315 days  
10 718 days  
11 161 days  
12 180 days  
13 564 days  
14 797 days

### 问题4（2分）：
我们原来的列名都是中文，现在我们想要把所有中文列名都改成英文。

1. 请根据给出的名称更改规则 rename_dic，对full_data的列名进行重命名，要求直接更新 full_data
2. 请列出重命名的这些列的数据，并用display和`head()`显示前几行

>**提示：** `df.rename()`更改列名，`dictionary.values()`可获得所有字典所有key对应的值, `list()`可以把其他对象转换成一个列表

In [33]:
# Rename rules as dictionary
rename_dic = {'姓名': 'Name',
             '性别': 'Gender',
             '年龄': 'Age',
             '投资账户': 'AC_No',
             '开户时间': 'AC_Open',
             '投资收益': 'Gain_Loss',
             '资产余额': 'AC_Balance'}
list(rename_dic.values())

['Name', 'Gender', 'Age', 'AC_No', 'AC_Open', 'Gain_Loss', 'AC_Balance']

In [53]:
# TO DO: Rename columns according to rename fules
# Then display renamed columns with head()
'''请输入你的代码'''
for key in rename_dic:
    full_data.rename(columns={key:rename_dic[key]}, inplace=True)
        
display(full_data.head(15))


Name Gender  Age 投资市场   AC_No    AC_Open  Gain_Loss  AC_Balance        new  \
0    张伟      男   32   A股  A00001 2016-03-02   30167.47   152582.83   张伟A00001   
1    王伟      男   26   A股  A00002 2016-06-04   87671.93   106538.72   王伟A00002   
2    王芳      女   50   A股  A00003 2017-04-21  -18090.58   125266.82   王芳A00003   
3    王芳      女   50   港股  H00003 2018-06-12   -7323.56   190558.09   王芳H00003   
4    李伟      男   31   A股  A00004 2016-09-01   27009.14   195731.90   李伟A00004   
5    李娜      女   58   美股  U00004 2018-04-24  -19278.16   132835.86   李娜U00004   
6   王秀英      女   56   A股  A00005 2017-07-05   40358.34   140649.78  王秀英A00005   
7   李秀英      女   45   A股  A00006 2018-06-23   36510.09   115750.13  李秀英A00006   
8   张秀英      女   52   A股  A00008 2017-11-03    6220.14   116819.85  张秀英A00008   
9    刘伟      男   46   A股  A00009 2017-10-21    7210.51   168387.54   刘伟A00009   
10   张敏      女   45   A股  A00010 2016-09-13   -3225.00   176235.81   张敏A00010   
11   李伟      男   29   A股  A00011 2018-03-24   42278.11   159249.38   李伟A00011   
12  李秀英      女   28   A股  A00012 2018-03-05   43597.13   121773.59  李秀英A00012   
13   张伟      男   32   港股  H00001 2017-02-14    2026.84   178320.86   张伟H00001   
14   王伟      男   26   港股  H00002 2016-06-26   54454.37   107915.45   王伟H00002   

      Aging  
0  913 days  
1  819 days  
2  498 days  
3   81 days  
4  730 days  
5  130 days  
6  423 days  
7   70 days  
8  302 days  
9  315 days  
10 718 days  
11 161 days  
12 180 days  
13 564 days  
14 797 days

### 问题5（2分）：
我们可以根据投资者过往的投资表现，把客户做一个简单的区分。对于那些投资收益好的客户，我们把他们认作“优秀投资者”

* 新增一列'type'，对于投资收益大于等于3万的客户，记录为'优秀投资者'，其余为'普通投资者'
* 并用display和`head()`显示 full_data 的前几行

>**要求：** 
1. 利用`df.loc[]`定位
2. 利用“`~`”符号对条件取非（条件：`full_data['Gain_Loss']>=30000`）

>**提示：**
- ~符号python常用的逻辑运算符之一，表示反运算。
- `df.loc['筛选条件', '新增列名'] = '' `<--通过这种方式可以赋值给新增列

In [61]:
# TO DO: Mark '优秀投资者' given Gain_Loss >=30000, otherwise mark '普通投资者'
'''请输入你的代码'''
full_data.loc[full_data['Gain_Loss'] >= 30000, 'type'] = '优秀投资者'
full_data.loc[~(full_data['Gain_Loss'] >= 30000), 'type'] = '普通投资者'
display(full_data.head(15))

Name Gender  Age 投资市场   AC_No    AC_Open  Gain_Loss  AC_Balance        new  \
0    张伟      男   32   A股  A00001 2016-03-02   30167.47   152582.83   张伟A00001   
1    王伟      男   26   A股  A00002 2016-06-04   87671.93   106538.72   王伟A00002   
2    王芳      女   50   A股  A00003 2017-04-21  -18090.58   125266.82   王芳A00003   
3    王芳      女   50   港股  H00003 2018-06-12   -7323.56   190558.09   王芳H00003   
4    李伟      男   31   A股  A00004 2016-09-01   27009.14   195731.90   李伟A00004   
5    李娜      女   58   美股  U00004 2018-04-24  -19278.16   132835.86   李娜U00004   
6   王秀英      女   56   A股  A00005 2017-07-05   40358.34   140649.78  王秀英A00005   
7   李秀英      女   45   A股  A00006 2018-06-23   36510.09   115750.13  李秀英A00006   
8   张秀英      女   52   A股  A00008 2017-11-03    6220.14   116819.85  张秀英A00008   
9    刘伟      男   46   A股  A00009 2017-10-21    7210.51   168387.54   刘伟A00009   
10   张敏      女   45   A股  A00010 2016-09-13   -3225.00   176235.81   张敏A00010   
11   李伟      男   29   A股  A00011 2018-03-24   42278.11   159249.38   李伟A00011   
12  李秀英      女   28   A股  A00012 2018-03-05   43597.13   121773.59  李秀英A00012   
13   张伟      男   32   港股  H00001 2017-02-14    2026.84   178320.86   张伟H00001   
14   王伟      男   26   港股  H00002 2016-06-26   54454.37   107915.45   王伟H00002   

      Aging   type  
0  913 days  优秀投资者  
1  819 days  优秀投资者  
2  498 days  普通投资者  
3   81 days  普通投资者  
4  730 days  普通投资者  
5  130 days  普通投资者  
6  423 days  优秀投资者  
7   70 days  优秀投资者  
8  302 days  普通投资者  
9  315 days  普通投资者  
10 718 days  普通投资者  
11 161 days  优秀投资者  
12 180 days  优秀投资者  
13 564 days  普通投资者  
14 797 days  优秀投资者

### 问题6（2分）：
对于上一步标注为'优秀投资者'的客户，我们来计算一下他们资产余额的平均值

* 计算'优秀投资者'的资产余额平均值，并且print出来。（仅一行代码）

>**提示：** `df.loc[]`定位

In [67]:
# TO DO: Calculate average AC_Balance
'''请输入你的代码'''
np.mean(full_data.loc[full_data['type']=='优秀投资者','AC_Balance'])

139057.1825

# 总结
我们通过以上的步骤，实践了一遍数据导入、数据内容观察和描述、简单的数据处理。

## 作业提交方法
> **注意：** 当你写完了所有的代码，并且回答了所有的问题。你就可以把你的 iPython Notebook 导出成 HTML 文件。你可以在菜单栏，这样导出**File -> Download as -> HTML (.html)**把这个 HTML 做为你的作业提交。

**提交方法一：**

可通过百度云上传，分享网盘链接

**提交方法二：**

可创建自己的github，分享项目链接

**提交方法三：**

登陆语雀并注册  https://www.yuque.com/   

选择右上角 + 号，新建知识库，任意命名。然后新建文档

在文档内，类似word的操作，插入附件，选择你导出的 HTML 文件。

右上角选择分享，开启分享，复制链接，把该链接在小程序里提交即可。